keps/sig-api-machinery/2876-crd-validation-expression-language/README.md

```diff
@@ -283,14 +283,31 @@ like the `all` macro, e.g. `self.all(listItem, <predicate>)` or `self.all(mapKey
     Strategy](https://kubernetes.io/docs/reference/using-api/server-side-apply/#merge-strategy) for
     details.
   - The use of "old" is congruent with how `AdmissionReview` identifies the existing object as
-    `oldObject`. To avoid name collisions `old<propertyName>` will be treated the same as a CEL
+    `oldSelf`. To avoid name collisions `old<propertyName>` will be treated the same as a CEL
     keyword for escaping purposes (see below).
   - xref [analysis of possible interactions with immutability and
     validation](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/1101-immutable-fields#openapi-extension-x-kubernetes-immutable).
 
-- If a object property collides with a CEL keyword (see RESERVED in [CEL Syntax](https://github.com/google/cel-spec/blob/master/doc/langdef.md#syntax)),
+- If a object property name is a CEL keyword (see RESERVED in [CEL Syntax](https://github.com/google/cel-spec/blob/master/doc/langdef.md#syntax)),
   it will be escaped by prepending a _ prefix. To prevent this from causing a subsequent collision, all properties with a `_` prefix will always be
   prefixed by `__` (generally, N+1 the existing number of `_`s).
+  
```

Comment:
are object property names which are not compatible with cel `IDENT` definition (e.g. containing `.`, `-`, ` ` characters) addressable? can they be accessed via indexing instead of field-access (e.g. `self["my.weird-property name"]`)?

Reply:
Oh noes. There is no way to access them. In CEL properties are only accessible via field selection (a.b), not by map indexing.

Apparently Firebase uses `__dot__` escaping.  Here's the schema I've prototyped:

- CEL keywords continue to escape to `_{keyword}` (and `_{keyword}` escaped to `__{keyword}`...)
- `.` escapes to `__dot__`
- `-` escapes to `__slash__`
- ` ` escapes to `__space__`
- `__` escapes to `__underscores__`

Note that a single `_` in the middle of identifier requires no escaping (so at least `field-foo` makes it through unescaped).

I am able to implement it (bidirectionally). If anyone has a better scheme let me know.

Reply:
Oh wait, are there any restrictions on what can be in a property name? I just tried a bunch of symbols and some unicode, and were able to use them all.

Reply:
There are a couple possible solutions here:

1) We introduce a general purpose escaping mechanism using the only symbol we have available: `_`.  There is not a lot of room to get creative, so it's probably going to be `__dot__` form for a small set of symbols that we give symbolic names to plus `__{unicode hex}__` for all the other non-alphanumeric characters.  I prototyped and fuzzed this and it can be made to work.  The `__{unicode hex}__` escaped characters are silly and totally impractical.

2) We pick a subset of characters that we're willing to support. If a property name has a character outside of that subset, we don't make it accessible in the validator rules.

3) We use CELs type integration to allow an object with properties to also support indexing, e.g. `spec["wacky field name"]`.  It looks like this might be possible to implement, but we'll loose static type checking for sure. CEL won't know at compile time what property type the map string key we're providing maps to.

I've committed #1 to anyone that is interested can have a look, but I'm not happy with the unicode escaping.

Reply:
Usually it's best to support the most limited API surface area like option 2.  Underscore should be supported as a valid identifier character.  However, CEL uses `.`, ` `, and `-` as delimiters and so they shouldn't be used within identifiers.

Many modern languages support unicode identifier characters, and I am not opposed to considering this if this is a common request. The alternative is a pattern for escaping field selections while preserving the more limited character set for identifiers. Supporting unicode feels like the more inclusive choice. Happy to discuss further.

Reply:
🤔 Is it possible to make a pass over the program and convert `spec["wacky field name"]` to `spec.wacky__space__field__space__name`. That is, convert the easier-for-human-to-spell format into the typesafe format? I suspect we would still have to explain the renaming rules, so maybe it doesn't reduce the complexity for users all that much.

This specific problem is probably something other CEL users (implementors like us) have had to deal with. I think I saw in one of the PRs that this underscore approach has been used elsewhere. Perhaps this problem and the alternative solutions we've investigated should be called out in the KEP?

Reply:
Per offline discussion we have gone with option 2 where we only make properties accessible to CEL if their names have alphanumeric, `_`, `.`, `-` or `/` characters (and don't start with a number).  We will use `__dot__` style escaping. I've updated this PR to match.